[PyTorch] Make c10::irange(x) generate the same assembly as for loop

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86841

`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

swolchok committed:
[PyTorch] Make c10::irange(x) generate the same assembly as for loop

`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

[ghstack-poisoned]

r-barnes commented on c10/test/util/irange_test.cpp:
```diff
@@ -6,6 +6,20 @@
 
+void f(int x);
+
+void iterateOldStyle(int bound) {
```
What is this being used for?

r-barnes commented on c10/test/util/irange_test.cpp:
```diff
@@ -6,6 +6,20 @@
+  }
+}
+
+void iterateIrange(int bound) {
```
Is this just to test compilation?

swolchok replied:
yep, can remove if you like

r-barnes commented on c10/util/irange.h:
```diff
@@ -54,20 +63,22 @@ struct integer_iterator : std::iterator<std::input_iterator_tag, I> {
 
 template <
     typename I,
+    bool one_sided = false,
```
I don't see a place where `one_sided` gets set to true?

r-barnes commented:
@swolchok - Thanks for this. Since `if constexpr` is part of C++17, we need to hold off on this until PyTorch completes its migration to C++17.

swolchok commented on c10/util/irange.h:
```diff
@@ -95,11 +106,8 @@ template <
-  // choosing the larger of {0, end} as the loop terminator
-  // Handles the case where end<0. irange only works for ranges >=0
-  return {Integer(), std::max(Integer(), end)};
+integer_range<Integer, true> irange(Integer end) {
```
here is where one_sided gets set to true

swolchok commented:
#85969 is the PR we are blocked on here.

swolchok commented on c10/util/irange.h:
```diff
@@ -39,11 +40,19 @@ struct integer_iterator : std::iterator<std::input_iterator_tag, I> {
 
   bool operator==(const integer_iterator& other) const {
-    return value == other.value;
+    if constexpr (one_sided) {
```
I think we can just make this an ordinary `if` statement and proceed.

swolchok committed:
Update on "[PyTorch] Make c10::irange(x) generate the same assembly as for loop"


`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

[ghstack-poisoned]

r-barnes commented on aten/src/ATen/native/TensorShape.cpp:
```diff
@@ -2889,7 +2889,7 @@ Tensor squeeze_qtensor(const Tensor& self, c10::optional<int64_t> dim) {
     auto axis = per_channel_quantizer->axis();
     int64_t shift = 0;
-    integer_range<int64_t> dims = dim.has_value() ? integer_range<int64_t>{dim.value(), dim.value() + 1} : c10::irange(self.dim());
+    integer_range<int64_t> dims = dim.has_value() ? integer_range<int64_t>{dim.value(), dim.value() + 1} : c10::irange(0, self.dim());
```
Is this a test artefact?

swolchok replied:
No, it's necessary for the two forks of the ?: operator to have the same type.

swolchok commented:
@pytorchbot merge -g

swolchok commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Make c10::irange(x) generate the same assembly as for loop"


`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make c10::irange(x) generate the same assembly as for loop"


`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge -g

swolchok commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Make c10::irange(x) generate the same assembly as for loop"


`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

[ghstack-poisoned]

swolchok committed:
Update on "[PyTorch] Make c10::irange(x) generate the same assembly as for loop"


`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge -g

swolchok commented:
@pytorchbot rebase

swolchok committed:
Update on "[PyTorch] Make c10::irange(x) generate the same assembly as for loop"


`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

[ghstack-poisoned]

swolchok commented:
@pytorchbot merge -g

swolchok committed:
Update on "[PyTorch] Make c10::irange(x) generate the same assembly as for loop"


`c10::irange(n)` generated an extra `sar` and `andn` instruction compared to a traditional `for` loop. now it doesn't.

Differential Revision: [D40321009](https://our.internmc.facebook.com/intern/diff/D40321009/)

[ghstack-poisoned]

malfet commented:
If you believe that failures are unrelated to the PR, feel free to merge it with `-g`.. But enabling C++17 across all our platforms feels a bit convoluted, but hopefully will get it done soon.

swolchok commented:
failure definitely looks unrelated

swolchok commented:
@pytorchbot merge -f

swolchok commented:
@pytorchbot merge -f 1 failure that sure looks unrelated

swolchok commented:
@pytorchbot merge -f '1 failure that sure looks unrelated'
